### ISSUE TITLE
Allow overriding via env vars custom pages/patches (just like config/cache)

### ISSUE THREAD

**Issue by eugenesvk** [dbrgn/tealdeer#306]
Currently you can override via Environment Variables
  - [Override Config Directory](https://dbrgn.github.io/tealdeer/config.html#override-config-directory) and 
  - [Override Cache Directory](https://dbrgn.github.io/tealdeer/config.html#override-cache-directory)
Similarly, I'd like to use env vars to override:
  - [Custom Pages](https://dbrgn.github.io/tealdeer/usage_custom_pages.html#custom-pages)
  - [Custom Patches](https://dbrgn.github.io/tealdeer/usage_custom_pages.html#custom-patches)

To allow using the same config file on various machines/users, since config only accepts absolute paths while it's easier to set relative paths in a shell config

**Comment by dbrgn** [comment]
Hmm, when setting relative paths, what would they be relative to?

Actually the `TEALDEER_CACHE_DIR` is something that is a leftover from when IIRC we didn't even have a config file. Since the config now contains a [`directories` section](https://dbrgn.github.io/tealdeer/config_directories.html), that's the place where I'd like to put all configuration values going forward. Adding additional env variables just makes things more complex. (I'd even like to deprecate the `TEALDEER_CACHE_DIR` variable and recommend using the config file instead, so that `TEALDEER_CONFIG_DIR` is the only env variable we need to check. What do you think, @niklasmohrin? We could still keep it working for quite some time in the code, but update docs.)

If absolute paths are a problem, then maybe that's where improvements could be made. Maybe supporting `~` expansion? (Not sure how that feels on Windows though, I think on Windows we'd need to support `%HOMEPATH%` instead 😕)

**Comment by eugenesvk** [comment]
> Hmm, when setting relative paths, what would they be relative to?

Relative to the global config path variable I've set up

> If absolute paths are a problem, then maybe that's where improvements could be made

Sure, and the least complicated way to improve it is so support env vars, where all of the issues like `~` and `%HOMEPATH` are already resolved in your shell config and only require a simple 
`$TEALDEER_CONFIG_DIR = f"{cfgH}/tldr"`
instead of forcing to use your resolver

**Comment by eugenesvk** [comment]
> cannot have all settings accessible only through env vars 

That's fine because env vars don't **replace** any configuration files, they only allow **overriding** it, so if users don't want to "clutter" their envs, then they can just... not set any env vars and use the configs

Also don't understand re. the performance hit — you're just reading from an environment var, there is no path resolution?

**Comment by niklasmohrin** [comment]
I suppose we could think about exposing all config settings through the environment. It would be a bit tricky to find the right spot to put this logic in, but could work. The performance hit I am talking about is the reading of lots of environment variables and running all the merging logic - I don't know what to expect there though

**Comment by eugenesvk** [comment]
Just to clarify — I'm not asking to expose all config settings, I don't see any point in that

**Comment by niklasmohrin** [comment]
Well, but only having these seems arbitrary again

**Comment by eugenesvk** [comment]
It's not arbitrary, these env solve the issue with relative paths, what issues does having everything solve???

**Comment by MatejKafka** [comment]
Just wanted to chime in – having environment variables for all paths is pretty useful for portable scenarios. With env vars, I can just write a wrapper script that sets the env vars to portable paths and invokes Tealdeer, and it works correctly when I move tealdeer to another path.

With the config file, I must instead patch the file whenever I move Tealdeer, which is much harder to do in a script. If the config file supported relative paths, I could set them once and then just set the single env var, so that would also work for me.

**Comment by niklasmohrin** [comment]
Yup agree, but as I said, I think we should come up with something unified for all options instead of adding exceptions for particular options now

**Comment by MatejKafka** [comment]
Thinking about it a bit more, imo the best solution would be to add support for config-relative paths in the config file, which solves the immediate issue with minimal implementation changes.

In the longer run, I agree that exposing all config file options in a more scriptable way might be worthwhile, but that's a much more elaborate change. As a sidenote, adding command-line parameters seems better to me than exposing everything using environment variables, since it is easier to use interactively and should have a bit less overhead.

**Comment by niklasmohrin** [comment]
Sounds good to me, it also seems straightforward to do: Just [`Path::join`](https://doc.rust-lang.org/std/path/struct.Path.html#method.join) the configured directory with the directory that the config file resides in. PRs (with tests) welcome

**Comment by jasonkarns** [comment]
I wanted to chime in here as well in support of this with a specific use case. I _think_ this usecase will be covered by #395 but as it's not released yet (I don't think), I wanted to confirm.

My current use case:
- I'm on macOS on multiple machines
- The multiple machines do not have the same username
- I use XDG dirs and have my ~/.config directory versioned with git to sync across machines
- I want to share & sync my tealdeer config (and custom pages) across machines

I'm currently _attempting_ to force tealdeer to respect XDG on mac (which works, though I would be very much in support of #311): 
```sh
export TEALDEER_CONFIG_DIR="${XDG_CONFIG_HOME:?}/tldr"
```

The problem, then, is that `directories.custom_pages_dir` requires an absolute path; which includes the `$USER` name (which varies across machines).

Will this issue (or #395) allow `~` to be resolved in the pathname? So that I can use:

```sh
$ cat $TEALDEER_CONFIG_DIR/config.toml 
[directories]
custom_pages_dir = "~/.config/tldr"
```

**Comment by niklasmohrin** [comment]
@jasonkarns None of the issues / PRs would add support for `~` in the config file I am afraid. However, #395 allows you to write something like
```toml
[directories]
custom_pages_dir = "../my-tealdeer-custom-pages"
```
which (in my case) results in
```shell
$ tldr --show-paths
Config dir:       /home/niklas/.config/tealdeer/ (OS convention)
Config path:      /home/niklas/.config/tealdeer/config.toml
Cache dir:        /home/niklas/.cache/tealdeer (OS convention)
Pages dir:        /home/niklas/.cache/tealdeer/tldr-pages/
Custom pages dir: /home/niklas/.config/tealdeer/../my-tealdeer-pages (config file)
```

In case that using `~` in the config file is a strongly desired feature, we should track it in a separate issue. For the time being, I believe that the above should work for your usecase though.

I suppose we could aim to implement #311 at some point too, since there is some demand for it. However, the next release of tealdeer will be mostly a cache-focused release (see https://github.com/tealdeer-rs/tealdeer/compare/v1.7.1...main and https://github.com/tealdeer-rs/tealdeer/milestone/6).

**Comment by jasonkarns** [comment]
Thanks @niklasmohrin . Yep, looks like #395 would provide a workaround for not expanding `~`, which covers my use case.

**Comment by MatejKafka** [comment]
Sounds good, it's very similar to git's `-c` parameter. I'd just call it `--config` (with a `-c` short name), `--extra-config` feels less discoverable.
